rust/lance/src/io/commit/conflict_resolver.rs

```diff
@@ -0,0 +1,722 @@
+    /// Has original fragment, plus a bool indicating whether a rewrite is needed.
+    initial_fragments: HashMap<u64, (Fragment, bool)>,
+    affected_rows: Option<&'a RowIdTreeMap>,
+}
+
+impl<'a> TransactionRebase<'a> {
+    pub async fn try_new(
+        dataset: &Dataset,
+        transaction: Transaction,
+        affected_rows: Option<&'a RowIdTreeMap>,
+    ) -> Result<Self> {
+        // We might not need to check for row-level conflicts anyways.
+        if !matches!(&transaction.operation,
+            Operation::Update { updated_fragments, .. } |
+            Operation::Delete { updated_fragments, .. }
+            if !updated_fragments.is_empty() && affected_rows.is_some(),
+        ) {
+            return Ok(Self {
+                transaction,
+                initial_fragments: HashMap::new(),
+                affected_rows: None,
+            });
+        }
+
+        log::info!(
+            "Manifest version: {}, transaction version: {}",
+            dataset.manifest.version,
+            transaction.read_version
+        );
+        let dataset = if dataset.manifest.version != transaction.read_version {
+            Cow::Owned(dataset.checkout_version(transaction.read_version).await?)
+        } else {
+            Cow::Borrowed(dataset)
+        };
+
+        let fragments = dataset.fragments().as_slice();
+
+        let modified_fragment_ids = transaction
+            .operation
+            .modified_fragment_ids()
+            .collect::<HashSet<_>>();
+
+        let initial_fragments = fragments
+            .iter()
+            .filter(|fragment| {
+                // Check if the fragment is modified by the transaction.
+                modified_fragment_ids.contains(&fragment.id)
+            })
+            .map(|fragment| (fragment.id, (fragment.clone(), false)))
+            .collect::<HashMap<_, _>>();
+
+        Ok(Self {
+            initial_fragments,
+            transaction,
+            affected_rows,
+        })
+    }
+
+    fn retryable_conflict_err(
+        &self,
+        other_transaction: &Transaction,
+        other_version: u64,
+        location: Location,
+    ) -> crate::Error {
+        crate::Error::RetryableCommitConflict {
+            version: other_version,
+            source: format!(
+                "This {} transaction was preempted by concurrent transaction {} at version {}. Please retry.",
+                self.transaction.operation, other_transaction.operation, other_version).into(),
+            location,
+        }
+    }
+
+    /// Check whether the transaction conflicts with another transaction.
+    ///
+    /// Will return an error if the transaction is not valid. Otherwise, it will
+    /// return Ok(()).
+    pub fn check_txn(
+        &mut self,
+        other_transaction: Option<&Transaction>,
+        other_version: u64,
+    ) -> Result<()> {
+        let Some(other_transaction) = other_transaction else {
```

Comment:
It's a little strange to accept `Option` and then immediately reject it if it is `None` though I could see how this might avoid a "err-if-none" check at all the callsites.

Reply:
yeah good point, all callers are doing `Some(other_transaction)` anyway, removed the Option.